README.md
# audit cookbook
[![Cookbook Version](http://img.shields.io/cookbook/v/audit.svg)][cookbook] [![Build Status](http://img.shields.io/travis/chef-cookbooks/audit.svg)][travis]

The `audit` cookbook allows you to run InSpec profiles as part of a Chef Client run. It downloads configured profiles from various sources like Chef Automate, Chef Supermarket or Git and reports audit runs to Chef Automate.

## Quickstart

The `audit` cookbook supports a number of different reporters and fetchers which can be confusing. Please see the [supported configurations documentation](https://github.com/chef-cookbooks/audit/blob/master/docs/supported_configuration.md) which has a few copy/paste examples to get you started quickly.

## Requirements

### Chef

- Chef Client >=12.20

### Support Matrix

#### Chef Automate

| Automate version   | InSpec version   | Audit Cookbook version   |
|--------------------|------------------|--------------------------|
|   < 0.8.0          |   ≤ 1.23.0       |   ≤ 3.1.0                |
|   ≥ 0.8.0          |   ≥ 1.24.0       |   ≥ 4.0.0                |
|   ≥ 2              |   ≥ 2.2.102      |   ≥ 7.1.0                |

#### Chef Infra Client

| Chef Client                | Audit Cookbook version    |
|----------------------------|---------------------------|
|   >= 15                    |   >= 8.0.0                |

Note:
When used with Chef Infra Client 15 and above, the Audit cookbook _must_ be >= 7.7.0. Otherwise, you will see the following failure. Of course, we recommend using the latest available Chef Infra Client and audit cookbook after testing in your non-production environments. Remove the `inspec_version` setting. See more detail in the [Inspec Gem Installation](#inspec-gem-installation) section

### Legacy Chef Infra Client Releases (pre-15.x) and Inspec

Newer Chef Infra Clients, in the 15.x and 16.x ranges ship with an
inspec-core gem that the audit cookbook in the proper version range
listed in the matrix above knows to use. It will not attempt to
download a gem in this case.

We find that 14.12 has the inspec-core 3.9.0 gem embedded which should
generate reports compatible with A2, but anything before that uses
InSpec gem v2.x which is NOT compatible with A2. To explain further,
Inspec 2.x will send reports to A2, but they will false positive, as
this report format is incompatible with A2

Chef Infra Client 14.x + plain, non inspec-core InSpec 3.9.3 installed by
the audit cookbook gem installer definitely works.  So in Chef Infra
Client 14.x before 14.12, we recommend this usage.

## Overview

### Component Architecture
```
 ┌──────────────────────┐    ┌──────────────────────┐    ┌─────────────────────┐
 │   Chef Infra Client  │    │   Chef Server Proxy  │    │    Chef Automate    │
 │                      │    │      (optional)      │    │                     │
 │ ┌──────────────────┐ │    │                      │    │                     │
 │ │                  │◀┼────┼──────────────────────┼────│  Profiles           │
 │ │  audit cookbook  │ │    │                      │    │                     │
 │ │                  │─┼────┼──────────────────────┼───▶│  Reports            │
 │ └──────────────────┘ │    │                      │    │                     │
 │                      │    │                      │    │                     │
 └──────────────────────┘    └──────────────────────┘    └─────────────────────┘
```

InSpec Profiles can be hosted from a variety of locations:

```
 ┌──────────────────────┐                                ┌─────────────────────┐
 │   Chef Infra Client  │     ┌───────────────────────┐  │    Chef Automate    │
 │                      │  ┌──│ Profiles(Supermarket, │  │                     │
 │ ┌──────────────────┐ │  │  │ GitHub, local, etc)   │  │                     │
 │ │                  │◀┼──┘  └───────────────────────┘  │                     │
 │ │  audit cookbook  │◀┼────────────────────────────────│  Profiles           │
 │ │                  │─┼───────────────────────────────▶│  Reports            │
 │ └──────────────────┘ │                                │                     │
 │                      │                                │                     │
 └──────────────────────┘                                └─────────────────────┘
```

## Usage

The audit cookbook needs to be configured for each node where the `chef-client` runs. The `audit` cookbook can be reused for all nodes, all node-specific configuration is done via Chef attributes.

### InSpec Gem Installation

**This section refers to EOL configuration. Starting with Chef Infra Client 15.x, only the embedded InSpec gem can be used and the `inspec_version` attribute will be ignored. Additionally, attempting to continue installing another version of the gem outside the Chef Infra Client inspec-core installation using the `inspec_gem()` resource can result in errors like the following. If you are running a Chef Infra Client release earlier than 15.x, we recommend testing an upgrade to Chef Infra Client 16.x, using the latest available audit cookbook version, and removing any `inspec_version` setting you may have done in your wrapper cookbook**

```
[2020-09-22T10:43:51-04:00] ERROR: Report handler Chef::Handler::AuditReport raised #<LoadError: cannot load such file -- chef-server/fetcher>
[2020-09-22T10:43:51-04:00] ERROR: /opt/chef/embedded/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
[2020-09-22T10:43:51-04:00] ERROR: /opt/chef/embedded/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
[2020-09-22T10:43:51-04:00] ERROR: /var/chef/cache/cookbooks/audit/files/default/handler/audit_report.rb:138:in `load_chef_fetcher'
[2020-09-22T10:43:51-04:00] ERROR: /var/chef/cache/cookbooks/audit/files/default/handler/audit_report.rb:62:in `report'
```

Beginning with version 3.x of the `audit` cookbook, the cookbook will first check to see if an InSpec gem is already installed. If it is, it will not attempt to install it. With the release of Chef Infra Client 14.2.x, the Chef omnibus package includes InSpec as a gem `inspec-core`. Recent releases of the audit cookbook use of this bundled component will reduce audit run times and also ensure that Chef users in air-gapped or firewalled environments can still use the `audit` cookbook without requiring gem mirrors, etc.

To install a different version of the InSpec gem, or to force installation of the gem, set the `node['audit']['inspec_version']` attribute to the version you wish to be installed.

Beginning with version 3.x of the `audit` cookbook, the default version of the InSpec gem to be installed (if it isn't already installed) is the latest version. Prior versions of the `audit` cookbook were version-locked to `inspec` version 1.15.0. **The use of Chef Infra Client versions 14.x and earlier was EOL as of April 30, 2020**

Note on AIX Support:

 * InSpec is only supported via the bundled InSpec gem shipped with version >= 14.2.x of the chef-client package.
 * Standalone InSpec gem installation or upgrade is not supported.
 * The default `nil` value of `node['audit']['inspec_version']` will ensure the above behavior is adhered to.

### Configure node

Once the cookbook is available in Chef Server, you need to add the `audit::default` recipe to the run-list of each node (or, preferably create a wrapper cookbook). The profiles are selected using the `node['audit']['profiles']` attribute. A list of example configurations are documented in [Supported Configurations](docs/supported_configuration.md). Below is another example demonstrating the different locations profiles can be "fetched" from:

```ruby
default['audit']['profiles']['linux-baseline'] = {
  'compliance': 'user/linux-baseline',
  'version': '2.1.0'
}

default['audit']['profiles']['ssh'] = {
  'supermarket': 'hardening/ssh-hardening'
}

default['audit']['profiles']['brewinc/win2012_audit'] = {
  'path': 'E:/profiles/win2012_audit'
}

default['audit']['profiles']['ssl'] = {
  'git': 'https://github.com/dev-sec/ssl-benchmark.git'
}

default['audit']['profiles']['ssh2'] = {
  'url': 'https://github.com/dev-sec/tests-ssh-hardening/archive/master.zip'
}
```

#### Attributes

You can also pass in [InSpec Attributes](https://www.inspec.io/docs/reference/profiles/) to your audit run. Do this by defining the attributes:

```ruby
default['audit']['attributes'] = {
  first_attribute: 'some value',
  second_attribute: 'another value',
}
```

#### Waivers

You can use Chef InSpec's [Waiver Feature](https://www.inspec.io/docs/reference/waivers/) to mark individual failing controls as being administratively accepted, either on a temporary or permanent basis. Prepare a waiver YAML file, and use your Chef Infra cookbooks to deliver the file to your converging node (for example, using [cookbook_file](https://docs.chef.io/resources/cookbook_file/) or [remote_file](https://docs.chef.io/resources/remote_file/)). Then set the attribute `default['audit']['waiver_file']` to the location of the waiver file on the local node, and Chef InSpec will apply the waivers.

### Reporting

#### Reporting to Chef Automate via Chef Server

To retrieve compliance profiles and report to Chef Automate through Chef Server, set the `reporter` and `profiles` attributes.

This requires Chef Client >= 12.16.42, Chef Server version 12.11.1, and Chef Automate 0.6.6 or newer, as well as integration between the Chef Server and Chef Automate. More details [here](https://docs.chef.io/automate/chef_infra_server/#connect-chef-infra-servers-to-chef-automate).

To upload profiles, you can use the [Automate API](https://docs.chef.io/automate/api/) or the `inspec compliance` subcommands (requires InSpec 1.7.2 or newer).

Attributes example of fetching from Automate, reporting to Automate both via Chef Server:

```ruby
default['audit']['reporter'] = 'chef-server-automate'
default['audit']['fetcher'] = 'chef-server'
default['audit']['profiles']['my-profile'] = {
  'compliance': 'john/my-profile'
}
```

#### Direct reporting to Chef Automate

To report directly to Chef Automate, set the `reporter` attribute to 'chef-automate' and specify where to fetch the `profiles` from.

* `insecure` - a `true` value will skip the SSL certificate verification. Default value is `false`

This method sends the report using the `data_collector.server_url` and `data_collector.token` options, defined in `client.rb`. It requires `inspec` version `0.27.1` or greater. Further information is available at Chef Docs: [Configure a Data Collector token in Chef Automate](https://docs.chef.io/automate/data_collection/)

```ruby
default['audit']['reporter'] = 'chef-automate'
default['audit']['profiles']['tmp_compliance_profile'] = {
  'url': 'https://github.com/nathenharvey/tmp_compliance_profile'
}
```

If you are using a self-signed certificate, please also read [how to add the Chef Automate certificate to the trusted_certs directory](https://docs.chef.io/automate/data_collection/#add-chef-automate-certificate-to-trusted_certs-directory)

Version compatibility matrix:

| Automate version   | InSpec version   | Audit Cookbook version   |
|--------------------|------------------|--------------------------|
|   < 0.8.0          |   ≤ 1.23.0       |   ≤ 3.1.0                |
|   ≥ 0.8.0          |   ≥ 1.24.0       |   ≥ 4.0.0                |


#### Compliance report size limitations

The size of the report being generated from running the compliance scan is influenced by a few factors like:
 * number of controls and tests in a profile
 * number of profile failures for the node
 * controls metadata (title, description, tags, etc)
 * number of resources (users, processes, etc) that are being tested

Depending on your setup, there are some limits you need to be aware of. A common one is Chef Server default (1MB) request size. Exceeding this limit will reject the report with `ERROR: 413 "Request Entity Too Large"`. For more details about these limits, please refer to [TROUBLESHOOTING.md](TROUBLESHOOTING.md#413-request-entity-too-large).

#### Write to file on disk

To write the report to a file on disk, simply set the `reporter` to 'json-file' like so:

```ruby
default['audit']['reporter'] = 'json-file'
default['audit']['profiles']['ssh2'] = {
  'path': '/some/base_ssh.tar.gz'
}
```

The resulting file will be written to `node['audit']['json_file']['location']` which defaults to
`<chef_cache_path>/cookbooks/audit/inspec-<YYYYMMDDHHMMSS>.json`. The path will also be output to
the Chef log:

```
[2017-08-29T00:22:10+00:00] INFO: Reporting to json-file
[2017-08-29T00:22:10+00:00] INFO: Writing report to /opt/kitchen/cache/cookbooks/audit/inspec-20170829002210.json
[2017-08-29T00:22:10+00:00] INFO: Report handlers complete
```

#### Enforce compliance with executed profiles

The `audit-enforcer` enables you to enforce compliance with executed profiles. If the system under test is determined to be non-compliant, this reporter will raise an error and fail the Chef Client run. To activate compliance enforcement, set the `reporter` attribute to 'audit-enforcer':

```ruby
default['audit']['reporter'] = 'audit-enforcer'
```

Note that detection of non-compliance will immediately terminate the Chef Client run. If you specify [multiple reporters](#multiple-reporters), place the `audit-enforcer` at the end of the list, allowing the other reporters to generate their output prior to run termination.

#### Multiple Reporters

To enable multiple reporters, simply define multiple reporters with all the necessary information
for each one.  For example, to report to Chef Automate and write to json file on disk:

```ruby
default['audit']['reporter'] = ['chef-server-automate', 'json-file']
default['audit']['profiles']['windows'] = {
  'compliance': 'base/windows'
}
)
```

### Profile Fetcher

#### Fetch profiles from Chef Automate via Chef Server

To enable reporting to Chef Automate with profiles from Chef Automate, you need to have Chef Server integrated with [Chef Automate](https://docs.chef.io/automate/data_collection/#configure-your-chef-infra-server-to-send-data-to-chef-automate). You can then set the `fetcher` attribute to 'chef-server'.

This allows the audit cookbook to fetch profiles stored in Chef Automate. For example:

```ruby
default['audit']['reporter'] = 'chef-server-automate'
default['audit']['fetcher'] = 'chef-server'
default['audit']['profiles']['ssh'] = {
  'compliance': 'base/ssh'
}
```

#### Fetch profiles directly from Chef Automate

This method fetches profiles using the `data_collector.server_url` and `data_collector.token` options, in `client.rb`. It requires `inspec` version `0.27.1` or greater. Further information is available at Chef Docs: [Configure a Data Collector token in Chef Automate](https://docs.chef.io/automate/data_collection/#setting-up-chef-infra-client-to-send-compliance-scan-data-directly-to-chef-automate)

```ruby
default['audit']['reporter'] = 'chef-automate'
default['audit']['fetcher'] = 'chef-automate'
default['audit']['profiles']['ssh'] = {
  'name': 'ssh',
}
```

## Interval Settings

If you have long running audit profiles that you don't wish to execute on every chef-client run,
you can enable an interval:

```
default['audit']['interval']['enabled'] = true
default['audit']['interval']['time'] = 1440 # once a day, the default value
```

The time attribute is in minutes.

You can enable the interval and set the interval time, along with your desired profiles,
 in an environment or role like this:

```json
  "audit": {
    "profiles": [
      {
        "name": "ssh",
        "compliance": "base/ssh"
      },
      {
        "name": "linux",
        "compliance": "base/linux"
      }
    ],
    "interval": {
      "enabled": true,
      "time": 1440
    }
  }
```

## Alternate Source Location for `inspec` Gem

If you are not able or do not wish to pull the `inspec` gem from rubygems.org,
you may specify an alternate source using:

```
# URI to alternate gem source (e.g. http://gems.server.com or filesytem location)
# root of location must host the *specs.4.8.gz source index
default['audit']['inspec_gem_source'] = 'http://internal.gem.server.com/gems'
```

Please note that all dependencies to the `inspec` gem must also be hosted in this location.

## Using Chef node data

While it is recommended that InSpec profiles should be self-contained and not rely on external data unless
necessary, there are valid use cases where a profile's test may exhibit different behavior depending on
aspects of the node under test.

There are two primary ways to pass Chef data to the InSpec run via the audit cookbook.

### Option 1: Explicitly pass necessary data (recommended)

Any data added to the `node['audit']['attributes']` hash will be passed as individual InSpec attributes.
This provides a clean interface between the Chef run and InSpec profile, allowing for easy assignment
of sane default values in the InSpec profile. This method is especially recommended if the InSpec profile
is expected to be used outside of the context of the audit cookbook so it's extra clear to profile
consumers what attributes are necessary.

In a wrapper cookbook or similar, set your Chef attributes:

```ruby
node.normal['audit']['attributes']['key1'] = 'value1'
node.normal['audit']['attributes']['debug_enabled'] = node['my_cookbook']['debug_enabled']
node.normal['audit']['attributes']['environment'] = node.chef_environment
```

... and then use them in your InSpec profile:

```ruby
environment = attribute('environment', description: 'The chef environment for the node', default: 'dev')

control 'debug-disabled-in-production' do
  title 'Debug logs disabled in production'
  desc 'Debug logs contain potentially sensitive information and should not be on in prod.'
  impact 1.0

  describe file('/path/to/my/app/config') do
    its('content') { should_not include "debug=true" }
  end

  only_if { environment == 'production' }
end
```

### Option 2: Use the chef node object

In the event where it is not practical to opt-in to pass certain attributes and data, the audit cookbook will
pass the Chef node object as an InSpec attribute named `chef_node`.

While this provides the ability to write more flexible profiles, it makes it more difficult to reuse profiles
outside of an audit cookbook run, requiring the profile user to know how to pass in a single attribute containing
Chef-like data. Therefore, it is recommended to use Option 1 whenever possible.

To use this option, first enable it in a wrapper cookbook or similar:

```ruby
node.override['audit']['chef_node_attribute_enabled'] = true
```

... and then use it in your profile:

```ruby
chef_node = attribute('chef_node', description: 'Chef Node')

control 'no-password-auth-in-prod' do
  title 'No Password Authentication in Production'
  desc 'Password authentication is allowed in all environments except production'
  impact 1.0

  describe sshd_config do
    its('PasswordAuthentication') { should cmp 'No' }
  end

  only_if { chef_node['chef_environment'] == 'production' }
end
```

## Using the InSpec Backend Cache

**Introduced in Audit Cookbook v6.0.0 and InSpec v1.47.0**

InSpec v1.47.0 provides the ability to cache the result of commands executed on the node being tested. This drastically improves InSpec performance when slower-running commands are run multiple times during execution.

This feature is **enabled by default** in the audit cookbook. If your profile runs a command multiple times and expects output to be different each time, you may have to disable this feature. To do so, set the `inspec_backend_cache` attribute to `false`:

```ruby
node.normal['audit']['inspec_backend_cache'] = false
```

## Troubleshooting

Please refer to [TROUBLESHOOTING.md](TROUBLESHOOTING.md).

Please let us know if you have any [issues](https://github.com/chef-cookbooks/audit/issues), we are happy to help.

## Run the tests for this cookbook:

Install [Chef Development Kit](https://downloads.chef.io/chefdk) on your machine.

```bash
# Install webmock gem needed by rspec
chef gem install webmock

# Run style checks
rake style


CHANGELOG.md
# Changelog

## Unreleased

- resolved cookstyle error: tasks/changelog.rake:48:25 convention: `Style/WordArray`
- resolved cookstyle error: tasks/changelog.rake:49:33 convention: `Style/WordArray`
- resolved cookstyle error: resources/inspec_gem.rb:1:1 refactor: `Chef/Deprecations/ResourceWithoutUnifiedTrue`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:40:7 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:56:7 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:72:7 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:88:16 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:116:16 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:137:16 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:159:16 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:181:16 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: spec/unit/recipes/default_spec.rb:198:16 warning: `ChefDeprecations/DeprecatedChefSpecPlatform`
- resolved cookstyle error: libraries/helper.rb:25:5 convention: `Style/RedundantAssignment`
- resolved cookstyle error: libraries/helper.rb:26:1 convention: `Layout/EmptyLinesAroundMethodBody`
- resolved cookstyle error: libraries/helper.rb:26:1 convention: `Layout/TrailingWhitespace`

## [v9.5.0](https://github.com/chef-cookbooks/audit/tree/v9.5.0) (2020-07-28)

[Full Changelog](https://github.com/chef-cookbooks/audit/compare/v9.4.0...v9.5.0)

**Closed issues:**

- Truncating large InSpec reports [\#422](https://github.com/chef-cookbooks/audit/issues/422)

**Merged pull requests:**

- Report even on inspec runtime exceptions [\#430](https://github.com/chef-cookbooks/audit/pull/430) ([alexpop](https://github.com/alexpop))
- Automated PR: Standardising Files [\#428](https://github.com/chef-cookbooks/audit/pull/428) ([xorimabot](https://github.com/xorimabot))

## [v9.4.0](https://github.com/chef-cookbooks/audit/tree/v9.4.0) (2020-05-20)

[Full Changelog](https://github.com/chef-cookbooks/audit/compare/v9.3.0...v9.4.0)

**Merged pull requests:**

- Release version 9.4.0 [\#427](https://github.com/chef-cookbooks/audit/pull/427) ([alexpop](https://github.com/alexpop))
- Truncate control results to max 50 by default [\#426](https://github.com/chef-cookbooks/audit/pull/426) ([alexpop](https://github.com/alexpop))

## [v9.3.0](https://github.com/chef-cookbooks/audit/tree/v9.3.0) (2020-05-15)

[Full Changelog](https://github.com/chef-cookbooks/audit/compare/v9.2.1...v9.3.0)

**Closed issues:**

- Audit cookbook is not upgrading in production mode [\#412](https://github.com/chef-cookbooks/audit/issues/412)
- Request: Update the Changelog [\#377](https://github.com/chef-cookbooks/audit/issues/377)

**Merged pull requests:**

- Release version 9.3.0 [\#425](https://github.com/chef-cookbooks/audit/pull/425) ([alexpop](https://github.com/alexpop))
- Add audit cookbook attributes support for the new InSpec options [\#424](https://github.com/chef-cookbooks/audit/pull/424) ([alexpop](https://github.com/alexpop))

## [v9.2.1](https://github.com/chef-cookbooks/audit/tree/v9.2.1) (2020-04-22)

[Full Changelog](https://github.com/chef-cookbooks/audit/compare/v9.2.0...v9.2.1)

**Closed issues:**

- missing waiver file fails all reporting [\#413](https://github.com/chef-cookbooks/audit/issues/413)

**Merged pull requests:**

- Release version 7.2.1 [\#421](https://github.com/chef-cookbooks/audit/pull/421) ([alexpop](https://github.com/alexpop))
- Ignore missing waiver files [\#420](https://github.com/chef-cookbooks/audit/pull/420) ([alexpop](https://github.com/alexpop))

## [v9.2.0](https://github.com/chef-cookbooks/audit/tree/v9.2.0) (2020-04-21)

[Full Changelog](https://github.com/chef-cookbooks/audit/compare/v9.1.0...v9.2.0)

**Closed issues:**

- Add Chef InSpec Waiver integration into Audit Cookbook [\#396](https://github.com/chef-cookbooks/audit/issues/396)

**Merged pull requests:**

- Release version 9.2.0 [\#419](https://github.com/chef-cookbooks/audit/pull/419) ([alexpop](https://github.com/alexpop))
- Support the Automate /compliance/profiles/metasearch endpoint to send reduced reports [\#418](https://github.com/chef-cookbooks/audit/pull/418) ([alexpop](https://github.com/alexpop))
- Update warnings on size limits, error on nil uuid [\#417](https://github.com/chef-cookbooks/audit/pull/417) ([btm](https://github.com/btm))
- Automated PR: Standardising Files [\#416](https://github.com/chef-cookbooks/audit/pull/416) ([xorimabot](https://github.com/xorimabot))
- Automated PR: Cookstyle Changes [\#415](https://github.com/chef-cookbooks/audit/pull/415) ([xorimabot](https://github.com/xorimabot))
- Error: The `Style/BracesAroundHashParameters` cop has been removed. [\#414](https://github.com/chef-cookbooks/audit/pull/414) ([Xorima](https://github.com/Xorima))
- Update changelog after release 9.1.0 [\#411](https://github.com/chef-cookbooks/audit/pull/411) ([alexpop](https://github.com/alexpop))

## [v9.1.0](https://github.com/chef-cookbooks/audit/tree/v9.1.0) (2020-03-03)

[Full Changelog](https://github.com/chef-cookbooks/audit/compare/v9.0.1...v9.1.0)

**Implemented enhancements:**

- Implement waiver file support [\#397](https://github.com/chef-cookbooks/audit/issues/397)
- Add waiver-file support [\#398](https://github.com/chef-cookbooks/audit/pull/398) ([clintoncwolfe](https://github.com/clintoncwolfe))

**Fixed bugs:**

- Send Report returned: 429 "Too Many Requests" from Automate Server [\#384](https://github.com/chef-cookbooks/audit/issues/384)

libraries/helper.rb
module ReportHelpers
  # Returns the uuid for the current converge
  def run_id
    return unless run_context &&
                  run_context.events &&
                  run_context.events.subscribers.is_a?(Array)
    run_context.events.subscribers.each do |sub|
      if sub.class == Chef::ResourceReporter && defined?(sub.run_id)
        return sub.run_id
      end
    end
    nil
  end

  # Returns the node's uuid
  def entity_uuid
    # the Chef::DataCollector::Messages API here is Chef < 15.0 backcompat and can be removed when Chef 14.x is no longer supported
    node['chef_guid'] || defined?(Chef::DataCollector::Messages) && Chef::DataCollector::Messages.node_uuid
  end

  # Convert the strings in the profile definitions into symbols for handling
  def tests_for_runner(profiles)
    profiles.map { |profile| Hash[profile.map { |k, v| [k.to_sym, v] }] }
  end

  def construct_url(server, path)
    # ensure we do not modify a frozen String
    srv = server.dup
    # sanitize inputs
    srv << '/' unless server =~ %r{/\z}
    path.sub!(%r{^/}, '')
    srv = URI(srv)
    srv.path = srv.path + path if path
    srv
  end

  def with_http_rescue(*)
    response = yield
    if response.respond_to?(:code)
      # handle non 200 error codes, they are not raised as Net::HTTPClientException
      handle_http_error_code(response.code) if response.code.to_i >= 300
    end
    response
  rescue Net::HTTPClientException => e
    Chef::Log.error e
    handle_http_error_code(e.response.code)
  end

  def handle_http_error_code(code)
    case code
    when /401|403/
      Chef::Log.error 'Auth issue: see audit cookbook TROUBLESHOOTING.md'
    when /404/
      Chef::Log.error 'Object does not exist on remote server.'
    when /413/
      Chef::Log.error 'You most likely hit the erchef request size in Chef Server that defaults to ~2MB. To increase this limit see audit cookbook TROUBLESHOOTING.md OR https://docs.chef.io/config_rb_server.html'
    when /429/
      Chef::Log.error "This error typically means the data sent was larger than Automate's limit (4 MB). Run InSpec locally to identify any controls producing large diffs."
    end
    msg = "Received HTTP error #{code}"
    Chef::Log.error msg
    raise msg if @raise_if_unreachable
  end

  def base_chef_server_url
    cs = chef_server_uri
    cs.path = ''
    cs.to_s
  end

  def chef_server_uri
    URI(Chef::Config.chef_server_url)
  end

  # used for interval timing
  def create_timestamp_file
    timestamp = Time.now.utc
    timestamp_file = File.new(report_timing_file, 'w')
    timestamp_file.puts(timestamp)
    timestamp_file.close
  end

  def report_timing_file
    # Will create and return the complete folder path for the chef cache location and the passed in value
    ::File.join(Chef::FileCache.create_cache_path('compliance'), 'report_timing.json')
  end

  def profile_overdue_to_run?(interval)
    # Calculate when a report was last created so we delay the next report if necessary
    return true unless ::File.exist?(report_timing_file)
    seconds_since_last_run = Time.now - ::File.mtime(report_timing_file)
    seconds_since_last_run > interval
  end

  def check_interval_settings(interval, interval_enabled, interval_time)
    # handle intervals
    interval_seconds = 0 # always run this by default, unless interval is defined
    if !interval.nil? && interval_enabled
      interval_seconds = interval_time * 60 # seconds in interval
      Chef::Log.debug "Auditing this machine every #{interval_seconds} seconds "
    end
    # returns true if profile is overdue to run
    profile_overdue_to_run?(interval_seconds)
  end

  # takes value of reporters and returns array to ensure backwards-compatibility
  def handle_reporters(reporters)
    return reporters if reporters.is_a? Array
    [reporters]
  end

  def cookbook_vendor_path
    File.expand_path('../files/default/vendor', __dir__)
  end

  def cookbook_handler_path
    File.expand_path('../files/default/handler', __dir__)
  end

  # Copies ['audit']['attributes'] into run_state for the audit_handler to read them later
  # Deletes ['audit']['attributes'] if instructed by ['audit']['attributes_save']
  def copy_audit_attributes
    node.run_state['audit_attributes'] = node['audit']['attributes']
    node.rm('audit', 'attributes') unless node['audit']['attributes_save']
  end

  def load_audit_handler
    libpath = ::File.join(cookbook_handler_path, 'audit_report')
    Chef::Log.info("loading handler from #{libpath}")
    $LOAD_PATH.unshift(libpath) unless $LOAD_PATH.include?(libpath)
    require libpath
    handler = Chef::Handler::AuditReport.new
    Chef::Config.send('report_handlers') << handler
    Chef::Config.send('exception_handlers') << handler
  end

  # taking node['audit'] as parameter so that it can be called from the chef-server fetcher as well
  # audit['collector'] is the legacy reporter,
  def get_reporters(audit)
    if audit.nil?
      Chef::Log.warn("node ['audit'] is not defined")
      return []
    end
    if audit['collector']
      Chef::Log.warn("node ['audit']['collector'] is deprecated and will be removed from the next major version of the cookbook. Please use node ['audit']['reporter']")
      return handle_reporters(audit['collector'])
    end
    handle_reporters(audit['reporter'])
  end

  # Extracts all the profile sha256 IDs from an inspec report
  def report_profile_sha256s(report)
    return [] unless report.is_a?(Hash) && report[:profiles].is_a?(Array)
    report[:profiles].map { |p| p[:sha256] }
  end

  # Strips profile metadata (title, copyright, controls title, code block, descriptions, etc)
  # from the `report` profiles that are not in the `missing_report_shas` array.
  # Control results are also stripped of the `start_time` and `run_time` if they don't reach the `run_time_limit` threshold (seconds)
  def strip_profiles_meta(report, missing_report_shas, run_time_limit)
    return report unless report.is_a?(Hash) && report[:profiles].is_a?(Array)
    report[:profiles].each do |p|
      next if missing_report_shas.include?(p[:sha256])
      # Profile 'name' is a required property. By not sending it in the report, we make it clear to the ingestion backend that the profile metadata has been stripped from this profile in the report.
      # Profile 'title' and 'version' are still kept for troubleshooting purposes in the backend.
      p.delete(:name)
      p.delete(:groups)
      p.delete(:copyright_email)
      p.delete(:copyright)
      p.delete(:summary)
      p.delete(:supports)
      p.delete(:license)
      p.delete(:maintainer)
      next unless p[:controls].is_a?(Array)
      p[:controls].each do |c|
        c.delete(:code)
        c.delete(:desc)
        c.delete(:descriptions)
        c.delete(:impact)
        c.delete(:refs)
        c.delete(:tags)
        c.delete(:title)
        c.delete(:source_location)
        c.delete(:waiver_data) if c[:waiver_data] == {}
        next unless c[:results].is_a?(Array)
        c[:results].each do |r|
          if r[:run_time].is_a?(Float) && r[:run_time] < run_time_limit
            r.delete(:start_time)
            r.delete(:run_time)
          end
        end
      end
    end
    report[:run_time_limit] = run_time_limit
    report
  end

  # Truncates the number of results per control in the report when they exceed max_results.
  # The truncation prioritizes failed and skipped results over passed ones.
  # Controls where results have been truncated will get a new object 'removed_results_counts'
  # with the status counts of the truncated results
  def truncate_controls_results(report, max_results)
    return report unless max_results.is_a?(Integer) && max_results > 0
    return report unless report.is_a?(Hash) && report[:profiles].is_a?(Array)
    report[:profiles].each do |profile|
      next unless profile[:controls].is_a?(Array)
      profile[:controls].each do |control|
        next unless control[:results].is_a?(Array)
        # Only bother with truncation if the number of results exceed max_results
        next unless control[:results].length > max_results
        res = control[:results]
        truncated = { failed: 0, skipped: 0, passed: 0 }
        res.sort_by! do |r|
          # Replacing "skipped" with "kipped" for the sort logic so that
          # the results are sorted in this order: failed, skipped, passed
          r[:status] == 'skipped' ? 'kipped' : r[:status]
        end
        # Count the results that will be truncated
        (max_results..res.length - 1).each do |i|
          case res[i][:status]
          when 'failed'
            truncated[:failed] += 1
          when 'skipped'
            truncated[:skipped] += 1
          when 'passed'
            truncated[:passed] += 1
          end
        end
        # Truncate the results array now
        control[:results] = res[0..max_results - 1]
        control[:removed_results_counts] = truncated
      end
    end
    report
  end

  # Contacts the metasearch Automate API to check which of the inspec profile sha256 ids
  # passed in via `report_shas` are missing from the Automate profiles metadata database.
  def missing_automate_profiles(automate_url, headers, report_shas)
    Chef::Log.debug "Checking the Automate profiles metadata for: #{report_shas}"
    meta_url = URI(automate_url)
    meta_url.path = '/compliance/profiles/metasearch'
    http = Chef::HTTP.new(meta_url.to_s)
    response_str = http.post(nil, "{\"sha256\": #{report_shas}}", headers)
    missing_shas = JSON.parse(response_str)['missing_sha256']
    unless missing_shas.empty?
      Chef::Log.info "Automate is missing metadata for the following profile ids: #{missing_shas}"
    end
    missing_shas
  rescue => e
    Chef::Log.error "missing_automate_profiles error: #{e.message}"
    # If we get an error it's safer to assume none of the profile shas exist in Automate
    report_shas
  end
end

::Chef::DSL::Recipe.include ReportHelpers
